src/protocol/ipa_prf/prf_sharding/mod.rs

```diff
@@ -953,64 +941,44 @@ pub mod tests {
-                capped_attributed_trigger_value.share_with(rng);
-
-            [
-                CappedAttributionOutputs {
-                    attributed_breakdown_key_bits: attributed_breakdown_key0,
-                    capped_attributed_trigger_value: capped_attributed_trigger_value0,
-                },
-                CappedAttributionOutputs {
-                    attributed_breakdown_key_bits: attributed_breakdown_key1,
-                    capped_attributed_trigger_value: capped_attributed_trigger_value1,
-                },
-                CappedAttributionOutputs {
-                    attributed_breakdown_key_bits: attributed_breakdown_key2,
-                    capped_attributed_trigger_value: capped_attributed_trigger_value2,
-                },
-            ]
-        }
-    }
-
-    impl Reconstruct<PreAggregationTestOutputInDecimal> for [&CappedAttributionOutputs; 3] {
+    impl<BK, TV> Reconstruct<PreAggregationTestOutputInDecimal>
+        for [&CappedAttributionOutputs<BK, TV>; 3]
+    where
+        BK: WeakSharedValue + CustomArray<Element = Boolean> + Field,
+        TV: WeakSharedValue + CustomArray<Element = Boolean> + Field,
+    {
         fn reconstruct(&self) -> PreAggregationTestOutputInDecimal {
             let [s0, s1, s2] = self;
-            let attributed_breakdown_key_bits: BitDecomposed<Gf2> = BitDecomposed::new(
-                s0.attributed_breakdown_key_bits
-                    .iter()
-                    .zip(s1.attributed_breakdown_key_bits.iter())
-                    .zip(s2.attributed_breakdown_key_bits.iter())
-                    .map(|((a, b), c)| [a, b, c].reconstruct()),
-            );
-            let capped_attributed_trigger_value_bits: BitDecomposed<Gf2> = BitDecomposed::new(
-                s0.capped_attributed_trigger_value
-                    .iter()
-                    .zip(s1.capped_attributed_trigger_value.iter())
-                    .zip(s2.capped_attributed_trigger_value.iter())
-                    .map(|((a, b), c)| [a, b, c].reconstruct()),
-            );
+            let bk_key_bits = [
+                s0.attributed_breakdown_key_bits.clone(),
+                s1.attributed_breakdown_key_bits.clone(),
+                s2.attributed_breakdown_key_bits.clone(),
+            ]
+            .reconstruct();
+            let capped_attributed_tv = [
+                s0.capped_attributed_trigger_value.clone(),
+                s1.capped_attributed_trigger_value.clone(),
+                s2.capped_attributed_trigger_value.clone(),
+            ]
+            .reconstruct();
+
+            let mut bd_key_decimal = 0_u128;
+            for i in 0..<BK as WeakSharedValue>::BITS {
+                bd_key_decimal += bk_key_bits.get(i.try_into().unwrap()).unwrap().as_u128() << i;
+            }
+
+            let mut tv_decimal = 0_u128;
+            for i in 0..<TV as WeakSharedValue>::BITS {
+                tv_decimal += capped_attributed_tv
```

Comment:
Don't you want to iterate over the reconstructed values like this:

```rust
let value = value_bits.into_iter().enumerate().fold(0, |b, (i, v)| b | u128::from(v) << i);
```
I would use "breakdown_key_bits" and "value_bits" for the variables too.  We already know that they are capped attribution outputs.  Similarly for the field names.
```

Reply:
Actually, I should just call "as_u128()" here and skip all of this nonsense.